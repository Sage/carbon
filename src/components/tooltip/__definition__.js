import Tooltip from './';
import Definition from './../../../demo/utils/definition';
import OptionsHelper from '../../utils/helpers/options-helper';

let definition = new Definition('tooltip', Tooltip, {
  description: `Additional information presented by hovering on a component.`,
  designerNotes: `
* A short paragraph of plain-text information shown to the user in context when they hover on any component. For example, to define how a particular figure is calculated, or the definition of a difficult technical term.
  `,
  relatedComponentsNotes: `
* Tooltip hovering on any Carbon icon? [Try Icon](/components/icon).
* Tooltip from a question mark icon? [Try Help](/components/help).
* Need a visual? [View Icons](/style/icons).__
`,
  type: 'miscellaneous',
  hiddenProps: [
    'id'
  ],
  propValues: {
    isVisible: true,
    children: "I'm a helpful tooltip that can display more information to a user."
  },
  propOptions: {
    align: OptionsHelper.alignAroundEdges,
    position: OptionsHelper.positions
  },
  propTypes: {
    align: 'String',
    children: 'Node',
    className: 'String',
    id: 'String',
    isVisible: 'Boolean',
    position: 'String'
  },
  propDescriptions: {
    align: 'The alignment of the pointer on the tooltip.',
    children: 'Child content to render in the tooltip.',
    className: 'Classes to be applied to the component.',
    id: 'The id attribute to apply to the tooltip',
    isVisible: 'Whether to show or hide the tooltip. Use this with a handler to hide and show the tooltip',
    position: 'The position of the tooltip relative to its target element.'
  }
});

export default definition;