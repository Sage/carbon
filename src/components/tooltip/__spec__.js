import React from 'react';
import TestUtils from 'react/lib/ReactTestUtils';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import Tooltip from './tooltip';

describe('tooltip', () => {
  let instance, alignedInstance, positionedInstance, hiddenInstance, wrapper;

  beforeEach(() => {
    instance = TestUtils.renderIntoDocument(
      <Tooltip isVisible={ true }>
        Some Helpful Content
      </Tooltip>
    );
    alignedInstance = TestUtils.renderIntoDocument(
      <Tooltip isVisible={ true } align='left'>
        Some Helpful Content
      </Tooltip>
    );

    positionedInstance = TestUtils.renderIntoDocument(
      <Tooltip isVisible={ true } position='top'>
        Some Helpful Content
      </Tooltip>
    );

    hiddenInstance = TestUtils.renderIntoDocument(
      <Tooltip isVisible={ false }>
        Some Helpful Content
      </Tooltip>
    );

    wrapper = shallow(
      <Tooltip isVisible={ true } data-element='bar' data-role='baz'>
        Some Helpful Content
      </Tooltip>
    );
  });

  describe('render', () => {
    let tooltip, alignedTooltip, positionedTooltip;

    beforeEach(() => {
      tooltip = ReactDOM.findDOMNode(instance);
      alignedTooltip = ReactDOM.findDOMNode(alignedInstance);
      positionedTooltip = ReactDOM.findDOMNode(positionedInstance);
    });

    describe('default render', () => {
      it('renders a content div', () => {
        expect(tooltip.classList[0]).toEqual('carbon-tooltip');
      });

      it('renders a pointer span', () => {
        let span = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-tooltip__pointer');
        expect(span.tagName).toEqual('SPAN');
      });

      it('aligns the pointer to center', () => {
        expect(tooltip.classList[2]).toEqual('carbon-tooltip--pointer-align-center');
      });

      it('positions the pointer on the top', () => {
        expect(tooltip.classList[1]).toEqual('carbon-tooltip--position-top');
      });
    });

    describe('when a prop of align is passed', () => {
      it('adds the appropriate class', () => {
        expect(alignedTooltip.classList[2]).toEqual('carbon-tooltip--pointer-align-left');
      });
    });

    describe('when a prop of position is passed', () => {
      it('adds the appropriate class', () => {
        expect(positionedTooltip.classList[1]).toEqual('carbon-tooltip--position-top');
      });
    });
  });

  describe('when the tooltip is toggled off', () => {
    it('does not render any content', () => {
      let hiddenTooltip = ReactDOM.findDOMNode(hiddenInstance);
      expect(hiddenTooltip).toEqual(null);
    });
  });

  describe('tags', () => {
    describe('on component', () => {
      it('includes correct component, element and role data tags', () => {
        window.RootTagTest.run(wrapper, 'tooltip', 'bar', 'baz');
      });
    });

    describe('on internal elements', () => {
      it("adds element tags to it's children", () => {
        window.ElementsTagTest.run(wrapper, ['container']);
      });
    });
  });
});
