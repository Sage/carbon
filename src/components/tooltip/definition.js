import Tooltip from './';
import Definition from './../../../demo/utils/definition';
import OptionsHelper from '../../utils/helpers/options-helper';

let definition = new Definition('tooltip', Tooltip, {
  description: `Additional information presented by hovering on a component.`,
  designerNotes: `
* A short paragraph of plain-text information shown to the user in context when they hover on any component. For example, to define how a particular figure is calculated, or the definition of a difficult technical term.
  `,
  relatedComponentsNotes: `
* Tooltip hovering on any Carbon icon? [Try Icon](/components/icon).
* Tooltip from a question mark icon? [Try Help](/components/help).
* Need a visual? [View Icons](/style/icons).__
`,
  type: 'miscellaneous',
  propValues: {
    isVisible: true,
    children: "I'm a helpful tooltip that can display more information to a user."
  },
  propOptions: {
    align: OptionsHelper.alignAroundEdges,
    position: OptionsHelper.positions
  },
  propTypes: {
    align: 'String',
    children: 'Node',
    position: 'String',
    isVisible: 'Boolean'
  },
  propDescriptions: {
    align: 'The alignment of the pointer on the tooltip.',
    children: 'Child content to render in the tooltip.',
    isVisible: 'Whether to show or hide the tooltip. Use this with a handler to hide and show the tooltip',
    position: 'The position of the tooltip relative to its target element.'
  }
});

export default definition;
