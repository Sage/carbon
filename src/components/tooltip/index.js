export { default } from './tooltip.component';
