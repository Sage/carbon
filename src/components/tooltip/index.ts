export { default } from "./tooltip.component";
export type { TooltipProps } from "./tooltip.component";
