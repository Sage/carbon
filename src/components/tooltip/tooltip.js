import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import tagComponent from '../../utils/helpers/tags';

/**
* A Tooltip widget.
*
* == How to use a Tooltip in a component:
*
* In your file:
*
*   import Tooltip from 'carbon-react/lib/components/tooltip'
*
* To render the Tooltip:
*
*   <Tooltip isVisible={ toggleTooltipHandler }>
*     My tooltip content
*   </Tooltip>
*
* You must pass a prop of 'isVisible' which is toggled to true or false.
*
* You can pass a prop of 'align' to the component which shifts the alignment of the pointer.
* This defaults to 'center'.
* You can also pass a prop of 'position' to the component which shifts the position of the pointer.
* This defaults to 'bottom'
*
* @class Tooltip
* @constructor
*/
class Tooltip extends React.Component {
  static propTypes = {
    /**
     * Sets alignment of pointer on tooltip
     *
     * Options: top, bottom, center, right, left
     *
     * @property align
     * @type {String}
     * @default 'center'
     */
    align: PropTypes.string,

    /**
     * Custom className
     *
     * @property className
     * @type {String}
     */
    className: PropTypes.string,

    /**
     * Children elements
     *
     * @property children
     * @type {Node}
     */
    children: PropTypes.node,

    /**
     * The id attribute to use for the tooltip
     *
     * @property id
     * @type {String}
     */
    id: PropTypes.string,

    /**
    * Whether to to show the Tooltip
    *
    * @property isVisible
    * @type {Boolean}
    * @default false
    */
    isVisible: PropTypes.bool,

    /**
     * Sets position of the tooltip
     *
     *
     * Options: top, bottom, right, left
     *
     * @property position
     * @type {String}
     * @default 'bottom'
     */
    position: PropTypes.string

  };

  static defaultProps = {
    align: 'center',
    position: 'top',
    className: '',
    isVisible: false
  };

  /**
   * Main classes
   *
   * @method mainClasses
   * @return {String} classNames for tooltip
   */
  get mainClasses() {
    return classNames(
      'carbon-tooltip',
      `carbon-tooltip--position-${this.props.position}`,
      `carbon-tooltip--pointer-align-${this.props.align}`,
      this.props.className
    );
  }

  /**
   * Return HTML for tooltip
   *
   * @method tooltipHTML
   * @return {JSX}
   */
  get tooltipHTML() {
    const contents = [
      this.props.children,
      <span key='pointer' className='carbon-tooltip__pointer' />
    ];

    const tooltipProps = {
      className: this.mainClasses,
      role: 'tooltip'
    };

    if (this.props.id) {
      tooltipProps.id = this.props.id;
    }

    return (
      <div
        { ...tooltipProps }
        { ...tagComponent('tooltip', this.props) }
      >
        <div className='carbon-tooltip__container'>
          { contents }
        </div>
      </div>
    );
  }

  /**
   * Renders the component.
   *
   * @method render
   */
  render() {
    if (!this.props.isVisible) { return null; }

    return this.tooltipHTML;
  }
}

export default Tooltip;
