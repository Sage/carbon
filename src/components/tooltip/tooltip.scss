@import 'colors';
@import 'general';
@import 'mixins';

.ui-tooltip {
  background-color: $grey-dark;
  border-radius: 4px;
  box-shadow: 0px 3px 10px 0px rgba(0,0,0,0.15);
  color: $white;
  max-width: 960px;
  min-width: 100px;
  padding: 10px 7px;
  position: absolute;
  text-align: center;
}

.ui-tooltip__pointer {
  height: 12px !important; // overrides the arrow mixin
  position: absolute;
}

.ui-tooltip__pointer--align-center.ui-tooltip__pointer--position-top {
  top: -15px;
  left: -webkit-calc(50% - 7px);
  left:    -moz-calc(50% - 7px);
  left:         calc(50% - 7px);

  &:before {
    @include arrow(up, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-center.ui-tooltip__pointer--position-bottom {
  left: -webkit-calc(50% - 7px);
  left:    -moz-calc(50% - 7px);
  left:         calc(50% - 7px);
  bottom: -12px;

  &:before {
    @include arrow(down, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-left.ui-tooltip__pointer--position-top {
  top: -15px;
  left: 4%;

  &:before {
    @include arrow(up, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-left.ui-tooltip__pointer--position-bottom {
  left: 4%;
  bottom: -12px;


  &:before {
    @include arrow(down, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-right.ui-tooltip__pointer--position-bottom {
  top: -15px;
  right: 7%;

  &:before {
    @include arrow(up, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-right.ui-tooltip__pointer--position-bottom {
  right: 7%;
  bottom: -12px;

  &:before {
    @include arrow(down, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-center.ui-tooltip__pointer--position-right {
  right: 0px;
  top: 33%;

  &:before {
    @include arrow(right, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-center.ui-tooltip__pointer--position-left {
  left: -15px;
  top: 33%;

  &:before {
    @include arrow(left, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-top.ui-tooltip__pointer--position-right {
  right: 0px;
  top: 10%;

  &:before {
    @include arrow(right, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-bottom.ui-tooltip__pointer--position-left {
  left: -15px;
  bottom: 10%;

  &:before {
    @include arrow(left, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-bottom.ui-tooltip__pointer--position-right {
  right: 0px;
  bottom: 10%;

  &:before {
    @include arrow(right, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}

.ui-tooltip__pointer--align-top.ui-tooltip__pointer--position-left {
  left: -15px;
  top: 10%;

  &:before {
    @include arrow(left, $app-tooltip-pointer-size, $grey-dark);
    position: absolute;
  }
}
