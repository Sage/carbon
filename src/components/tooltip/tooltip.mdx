import { Meta, ArgTypes, Canvas } from "@storybook/blocks";

import * as TooltipStories from "./tooltip.stories";

<Meta title="Tooltip" of={TooltipStories} />

# Tooltip

<a
  target="_blank"
  href="https://zeroheight.com/2ccf2b601/p/93ac95-tooltip"
  style={{ color: "#007E45", fontWeight: "bold", textDecoration: "underline" }}
  rel="noreferrer"
>
  Product Design System component
</a>

The `Tooltip` component is built using `tippyjs` to support features such as dynamic positioning where the tooltip will
flip if there is no space in its default placement; and the arrow and tooltip will track the target element when a user scrolls
or the window resizes.

## Contents

- [Quick Start](#quick-start)
- [Examples](#examples)
- [Props](#props)

## Quick Start

```javascript
import Tooltip from "carbon-react/lib/components/tooltip";
```

## Examples

### Default

To use the Tooltip, pass in the component you want it to display on in via the `children` prop, any component passed in
must support `forwardRef`. By default the Tooltip does not need to be controlled and will display when the target element
is focused or hovered. The default `position` is "top" and `size` is "M". The `message` prop is the text displayed in the
tooltip and it and the `children` prop are required.

<Canvas of={TooltipStories.Default} />

### Controlled

The Tooltip can also be controlled via the `isVisible` prop.

<Canvas of={TooltipStories.Controlled} />

### Positioning

The Tooltip supports four `position`s: "top", "bottom", "left" and "right". However, if there is no room for it to display
in this position it will attempt to flip and dynamically place itself in the opposite position. The Tooltip will also track
the target element when a user scrolls. The flipping and tracking features can be seen by adjusting the window size
(ideally in the canvas tab) and scrolling.

<Canvas of={TooltipStories.Positioning} />

### Overriding the default flip behaviour

By default if there is no room for the Tooltip to display in the given position it will attempt to flip and dynamically
place itself in the opposite position. You can pass an array of `flipOverrides` which will alter the behaviour from the
default, the Tooltip will flip to the positions defined in the array in the order they are defined. The overriding
can be seen by adjusting the window size (ideally in the canvas tab) and scrolling. The example should render the
Tooltip "bottom" intitially, then flip to "right" when there is not enough room below, and finally to the "left" position
when there is no space to render to the right anymore.

<Canvas of={TooltipStories.FlipBehaviourOverrides} />

### Large tooltip

<Canvas of={TooltipStories.LargeTooltip} />

### Types

The `Tooltip` can be rendered as two different `type`s which will alter the colour of the background: "error"
or if no value is passed it will default to a black background.

<Canvas of={TooltipStories.Types} />

### Color overrides

The `Tooltip` background and font color can be overridden via the `bgColor` and `fontColor` props and accept any valid
css value. Any values passed to the `bgColor` prop will override the color applied by the `type`

<Canvas of={TooltipStories.ColorOverrides} />

## Props

### Tooltip

<ArgTypes of={TooltipStories} />
