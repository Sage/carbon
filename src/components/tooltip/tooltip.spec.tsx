import React from "react";
import { mount, shallow } from "enzyme";

import Tooltip, { TooltipProps, InputSizes } from "./tooltip.component";
import StyledTooltipWrapper from "./tooltip.style";
import StyledTooltipPointer from "./tooltip-pointer.style";
import { TooltipPositions } from "./tooltip.config";

import { assertStyleMatch } from "../../__spec_helper__/test-utils";
import CarbonScopedTokensProvider from "../../style/design-tokens/carbon-scoped-tokens-provider/carbon-scoped-tokens-provider.component";

const positions: TooltipPositions[] = ["top", "bottom", "left", "right"];

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function render(props: Partial<TooltipProps> = {}, renderer: any = mount) {
  const children = <div>foo</div>;
  const message = "foo";

  return renderer(
    <Tooltip {...props} message={message || props.message}>
      {props.children || children}
    </Tooltip>
  );
}

describe("Tooltip", () => {
  describe("default props", () => {
    it("matches snapshot", () => {
      const wrapper = render({}, shallow);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("controlled", () => {
    it("matches snapshot when isVisible is true", () => {
      const wrapper = render({ isVisible: true }, shallow);

      expect(wrapper).toMatchSnapshot();
    });

    it("matches snapshot when isVisible is false", () => {
      const wrapper = render({ isVisible: false }, shallow);

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe("styles", () => {
    describe("TooltipWrapper", () => {
      describe("default", () => {
        it("applies the default styles", () => {
          assertStyleMatch(
            {
              bottom: "auto",
              right: "auto",
              position: "relative",
              maxWidth: "300px",
              zIndex: "6000",
              textAlign: "left",
              color: "var(--colorsSemanticNeutralYang100)",
              display: "inline-block",
              padding: "8px 12px",
              wordBreak: "break-word",
              whiteSpace: "pre-wrap",
              fontSize: "14px",
              lineHeight: "1.5rem",
              fontWeight: "400",
              backgroundColor: "var(--colorsSemanticNeutral500)",
            },
            render().find(StyledTooltipWrapper)
          );
        });

        it("applies the correct styles when size is 'large'", () => {
          assertStyleMatch(
            { fontSize: "16px" },
            render({ size: "large" }).find(StyledTooltipWrapper)
          );
        });

        it("applies the correct styles  type === 'error'", () => {
          assertStyleMatch(
            { backgroundColor: "var(--colorsSemanticNegative500)" },
            render({ type: "error" }).find(StyledTooltipWrapper)
          );
        });

        it.each(positions)(
          "applies the offset when position is %s",
          (position) => {
            assertStyleMatch(
              { [position]: "1px" },
              render({ position }).find(StyledTooltipWrapper)
            );
          }
        );
      });

      describe("color props", () => {
        it("overrides default background when a valid css string is passed via 'bgColor'", () => {
          assertStyleMatch(
            { backgroundColor: "pink" },
            render({ bgColor: "pink" }).find(StyledTooltipWrapper)
          );

          assertStyleMatch(
            { borderTop: "8px solid pink" },
            render({ bgColor: "pink" }).find(StyledTooltipPointer)
          );
        });

        it("overrides the type prop if background color passed via 'bgColor'", () => {
          assertStyleMatch(
            { backgroundColor: "pink" },
            render({ type: "error", bgColor: "pink" }).find(
              StyledTooltipWrapper
            )
          );

          assertStyleMatch(
            { borderTop: "8px solid pink" },
            render({ type: "error", bgColor: "pink" }).find(
              StyledTooltipPointer
            )
          );
        });

        it("overrides default font color when a valid css string is passed via 'fontColor'", () => {
          assertStyleMatch(
            { color: "pink" },
            render({ fontColor: "pink" }).find(StyledTooltipWrapper)
          );
        });
      });
    });

    describe("when the tooltip targets a component that is a part of an input", () => {
      const isTopOrBottom = (position: TooltipPositions) =>
        ["top", "bottom"].includes(position);
      const offsets = (
        position: TooltipPositions
      ): Record<InputSizes, string> => ({
        small: "5px",
        medium: isTopOrBottom(position) ? "4px" : "2px",
        large: isTopOrBottom(position) ? "0px" : "-2px",
      });

      describe.each(positions)("and position is %s", (position) => {
        it.each<InputSizes>(["small", "medium", "large"])(
          "sets the offset as expected when size is %s",
          (size) => {
            assertStyleMatch(
              {
                [position]: offsets(position)[size],
              },
              render({ position, isPartOfInput: true, inputSize: size }).find(
                StyledTooltipWrapper
              )
            );
          }
        );
      });
    });

    describe("TooltipPointer", () => {
      describe("when position === 'top' / default", () => {
        it("applies the correct styles", () => {
          assertStyleMatch(
            {
              position: "absolute",
              width: "0",
              height: "0",
              borderLeft: "8px solid transparent",
              borderRight: "8px solid transparent",
              borderTop: "8px solid var(--colorsSemanticNeutral500)",
              bottom: "calc(-1 * var(--spacing100))",
            },
            render().find(StyledTooltipPointer)
          );
        });
      });

      it("applies the correct styles when type === 'error'", () => {
        assertStyleMatch(
          { borderTop: "8px solid var(--colorsSemanticNegative500)" },
          render({ type: "error" }).find(StyledTooltipPointer)
        );
      });

      describe('when position === "bottom"', () => {
        it("applies the correct styles", () => {
          assertStyleMatch(
            {
              borderLeft: "8px solid transparent",
              borderRight: "8px solid transparent",
              borderBottom: "8px solid var(--colorsSemanticNeutral500)",
              top: "calc(-1 * var(--spacing100))",
            },
            render({ position: "bottom" }).find(StyledTooltipPointer)
          );
        });

        it("applies the correct styles when type === 'error'", () => {
          assertStyleMatch(
            { borderBottom: "8px solid var(--colorsSemanticNegative500)" },
            render({ position: "bottom", type: "error" }).find(
              StyledTooltipPointer
            )
          );
        });
      });

      describe('when position === "left"', () => {
        it("applies the correct styles", () => {
          assertStyleMatch(
            {
              borderTop: "8px solid transparent",
              borderBottom: "8px solid transparent",
              borderLeft: "8px solid var(--colorsSemanticNeutral500)",
              right: "calc(-1 * var(--spacing100))",
            },
            render({ position: "left" }).find(StyledTooltipPointer)
          );
        });

        it("applies the correct styles when type === 'error'", () => {
          assertStyleMatch(
            { borderLeft: "8px solid var(--colorsSemanticNegative500)" },
            render({ position: "left", type: "error" }).find(
              StyledTooltipPointer
            )
          );
        });
      });

      describe('when position === "right"', () => {
        it("applies the correct styles", () => {
          assertStyleMatch(
            {
              borderTop: "8px solid transparent",
              borderBottom: "8px solid transparent",
              borderRight: "8px solid var(--colorsSemanticNeutral500)",
              left: "calc(-1 * var(--spacing100))",
            },
            render({ position: "right" }).find(StyledTooltipPointer)
          );
        });

        it("applies the correct styles when type === 'error'", () => {
          assertStyleMatch(
            { borderRight: "8px solid var(--colorsSemanticNegative500)" },
            render({ position: "right", type: "error" }).find(
              StyledTooltipPointer
            )
          );
        });
      });
    });
  });

  describe("flipOverrides", () => {
    let mockGlobal: jest.SpyInstance;

    beforeEach(() => {
      mockGlobal = jest
        .spyOn(global.console, "error")
        .mockImplementation(() => undefined);
    });

    afterEach(() => {
      mockGlobal.mockReset();
    });

    it("does not throw an error if a valid array is passed", () => {
      render({ flipOverrides: ["top", "bottom"] });

      // eslint-disable-next-line no-console
      expect(console.error).not.toHaveBeenCalled();
    });

    it("throws an error if a invalid array is passed", () => {
      expect(() => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore:next-line
        render({ flipOverrides: ["foo", "bar"] });
      }).toThrow();
    });
  });

  describe("Design tokens", () => {
    it("wraps content with CarbonScopedTokensProvider", () => {
      const wrapper = render();

      const carbonScopedTokensProvider = wrapper.find(
        CarbonScopedTokensProvider
      );

      expect(
        carbonScopedTokensProvider.find(StyledTooltipWrapper).exists()
      ).toBe(true);
    });
  });

  describe("Ref forwarding", () => {
    it("should forward a ref object correctly", () => {
      const testRef = { current: null };

      const wrapper = mount(
        <Tooltip message="foo" isVisible ref={testRef}>
          <span>Test tooltip</span>
        </Tooltip>
      );

      wrapper.update();

      expect(testRef.current).toEqual(
        wrapper.find(StyledTooltipWrapper).getDOMNode()
      );
    });

    it("should forward a callback ref correctly", () => {
      const testCallbackRef = jest.fn();

      const wrapper = mount(
        <Tooltip message="foo" isVisible ref={testCallbackRef}>
          <span>Test tooltip</span>
        </Tooltip>
      );

      wrapper.update();

      expect(testCallbackRef).toHaveBeenCalledWith(
        wrapper.find(StyledTooltipWrapper).getDOMNode()
      );
    });
  });
});
