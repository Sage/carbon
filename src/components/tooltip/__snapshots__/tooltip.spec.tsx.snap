// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip controlled matches snapshot when isVisible is false 1`] = `
<ForwardRef(TippyWrapper)
  delay={100}
  placement="top"
  plugins={
    Array [
      Object {
        "defaultValue": false,
        "fn": [Function],
        "name": "sticky",
      },
    ]
  }
  popperOptions={
    Object {
      "modifiers": Array [
        Object {
          "name": "computeStyles",
          "options": Object {
            "gpuAcceleration": false,
          },
        },
      ],
    }
  }
  render={[Function]}
  sticky={true}
  visible={false}
>
  <div>
    foo
  </div>
</ForwardRef(TippyWrapper)>
`;

exports[`Tooltip controlled matches snapshot when isVisible is true 1`] = `
<ForwardRef(TippyWrapper)
  delay={100}
  placement="top"
  plugins={
    Array [
      Object {
        "defaultValue": false,
        "fn": [Function],
        "name": "sticky",
      },
    ]
  }
  popperOptions={
    Object {
      "modifiers": Array [
        Object {
          "name": "computeStyles",
          "options": Object {
            "gpuAcceleration": false,
          },
        },
      ],
    }
  }
  render={[Function]}
  sticky={true}
  visible={true}
>
  <div>
    foo
  </div>
</ForwardRef(TippyWrapper)>
`;

exports[`Tooltip default props matches snapshot 1`] = `
<ForwardRef(TippyWrapper)
  delay={100}
  placement="top"
  plugins={
    Array [
      Object {
        "defaultValue": false,
        "fn": [Function],
        "name": "sticky",
      },
    ]
  }
  popperOptions={
    Object {
      "modifiers": Array [
        Object {
          "name": "computeStyles",
          "options": Object {
            "gpuAcceleration": false,
          },
        },
      ],
    }
  }
  render={[Function]}
  sticky={true}
>
  <div>
    foo
  </div>
</ForwardRef(TippyWrapper)>
`;
